include/albatross/src/linalg/block_diagonal.hpp

```diff
@@ -102,6 +126,55 @@ inline Eigen::Matrix<_Scalar, _Rows, _Cols> BlockDiagonalLDLT::sqrt_solve(
   return output;
 }
 
+template <class _Scalar, int _Options, typename _StorageIndex>
+inline Eigen::Matrix<_Scalar, Eigen::Dynamic, Eigen::Dynamic>
+BlockDiagonalLDLT::solve(
+    const Eigen::SparseMatrix<_Scalar, _Options, _StorageIndex> &rhs) const {
+  ALBATROSS_ASSERT(cols() == rhs.rows());
+  Eigen::Index i = 0;
+  Eigen::Matrix<_Scalar, Eigen::Dynamic, Eigen::Dynamic> output(rows(),
+                                                                rhs.cols());
+  for (const auto &b : blocks) {
+    const auto rhs_chunk = rhs.block(i, 0, b.rows(), rhs.cols());
+    output.block(i, 0, b.rows(), rhs.cols()) = b.solve(rhs_chunk);
+    i += b.rows();
+  }
+  return output;
+}
+
+template <typename Op, typename LhsType, typename RhsType>
+inline Eigen::MatrixXd BlockDiagonalLDLT::solve(
+    const Eigen::CwiseBinaryOp<Op, LhsType, RhsType> &rhs_orig) const {
```

Comment:
What happens at the moment if you do:
```
Eigen::MatrixXd a, b;
block_diag.solve(a.cwiseProduct(b));
```
I would have guessed it'd find the `solve(Matrix<_Scalar, _Rows, _Cols>)` call [here](https://github.com/swift-nav/albatross/blob/master/include/albatross/src/linalg/block_diagonal.hpp#L29) and do effectively the same thing as these new methods?

What if you change `Matrix<...>` to `MatrixBase<...>`?

Reply:
`MatrixBase<Derived>` doesn't cover `CWiseBinaryOp` for sure.  But I think it could cover `Block`.

Reply:
There are a couple problems here.  One is that we have to declare a type for `output`.  So if we use `MatrixBase<Derived>`, then if you pass it a `CWiseBinaryOp`, it tries to declare one and assign to it.  This is clearly no good as `CWiseBinaryOp`s are never lvalues.  But how else can we determine the return type so as to preserve sparsity?  We don't care about that for `rcond()`, but we do care about it (I think) in PIC.

Reply:
The Right Thing to do here is to make `BlockDiagonalLDLT` cooperate with the Eigen expression system.

I think I'm going to just make it return `MatrixXd` for now.  But also I was wrong above that `MatrixBase` covers `Block`.

Reply:
This change is up, which allows removing only the overload for `CWiseBinaryOp`.